blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitReadSaveService.java

```diff
@@ -0,0 +1,286 @@
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+        if (false) {
+            return new GitReadSaveRequest(
+                StringUtils.defaultIfEmpty(readUrl, existingReadUrl),
+                StringUtils.defaultIfEmpty(writeUrl, StringUtils.defaultIfEmpty(existingWriteUrl, readUrl)),
+                StringUtils.defaultIfEmpty(readCredentialId, existingReadCredentialId),
+                StringUtils.defaultIfEmpty(writeCredentialId, existingWriteCredentialId),
+                StringUtils.defaultIfEmpty(branch, defaultBranch),
+                commitMessage,
+                StringUtils.defaultIfEmpty(sourceBranch, defaultBranch),
+                filePath,
+                contents,
+                gitTool,
+                gitSource
+            );
+        }
+        return new GitSCMReadSaveService.GitSCMReadSaveRequest(
+            item,
+            StringUtils.defaultIfEmpty(readUrl, existingReadUrl),
+            StringUtils.defaultIfEmpty(writeUrl, StringUtils.defaultIfEmpty(existingWriteUrl, readUrl)),
+            StringUtils.defaultIfEmpty(readCredentialId, existingReadCredentialId),
+            StringUtils.defaultIfEmpty(writeCredentialId, existingWriteCredentialId),
+            StringUtils.defaultIfEmpty(branch, defaultBranch),
+            commitMessage,
+            StringUtils.defaultIfEmpty(sourceBranch, defaultBranch),
+            filePath,
+            contents,
+            gitTool,
+            gitSource
+        );
+    }
+
+    private GitReadSaveRequest makeSaveRequest(Item item, StaplerRequest req) throws IOException, InterruptedException {
+        String readUrl = req.getParameter("readUrl");
+        return makeSaveRequest(
+            item,
+            readUrl,
+            req.getParameter("writeUrl"),
+            req.getParameter("readCredentialId"),
+            req.getParameter("writeCredentialId"),
+            req.getParameter("branch"),
+            req.getParameter("commitMessage"),
+            req.getParameter("sourceBranch"),
+            req.getParameter("path"),
+            Base64.decode(req.getParameter("contents"))
+        );
+    }
+
+    private GitReadSaveRequest makeSaveRequest(Item item, JSONObject json) throws IOException, InterruptedException {
+        JSONObject content = json.getJSONObject("content");
+        return makeSaveRequest(
+            item,
+            null,
+            null,
+            null,
+            null,
+            content.getString("branch"),
+            content.getString("message"),
+            content.getString("sourceBranch"),
+            content.getString("path"),
+            Base64.decode(content.getString("base64Data"))
+        );
+    }
+
+    @Override
+    public Object getContent(StaplerRequest req, Item item) {
+        GitReadSaveRequest r = null;
+        try {
+            r = makeSaveRequest(item, req);
+            r.cloneRepo();
```

Comment:
So every time we load editor we are cloning the repo? Git pipeline might point to a remote Github/GHE repo as well, so that might be expensive cloning every time? Maybe worth checking with @stephenc, he implemented shallow cloning with github/bitbucket and maybe Git.